test/e2e/upgrade_test.go

```diff
@@ -0,0 +1,170 @@
+// them on version annotation mismatch.
+func testUpgradeVersion(t *testing.T) {
+	testCtx, err := NewTestContext(t)
+	require.NoError(t, err)
+
+	operatorVersion, err := getWMCOVersion()
+	require.NoError(t, err, "could not get WMCO version")
+
+	err = testCtx.waitForWindowsNodes(gc.numberOfNodes, true, false, true)
+	if err != nil {
+		t.Fatalf("windows node upgrade failed  with %v", err)
+	}
+
+	// Get all the Windows nodes
+	nodes, err := framework.Global.KubeClient.CoreV1().Nodes().List(context.TODO(),
+		metav1.ListOptions{LabelSelector: nc.WindowsOSLabel})
+	require.NoError(t, err)
+
+	for _, node := range nodes.Items {
+		t.Run(node.GetName(), func(t *testing.T) {
+			require.Containsf(t, node.Annotations, nc.VersionAnnotation, "node missing version annotation")
+			assert.Equal(t, operatorVersion, node.Annotations[nc.VersionAnnotation], "Node recreation failed. WMCO version annotation mismatch")
+		})
+	}
+}
+
+// testTamperAnnotation tests if the operator deletes machines and recreates them, if the node annotation is changed to an invalid value
+// with the expected annotation when the operator is in running state
+func testTamperAnnotation(t *testing.T) {
+	testCtx, err := NewTestContext(t)
+	require.NoError(t, err)
+
+	operatorVersion, err := getWMCOVersion()
+	require.NoError(t, err, "could not get WMCO version")
+
+	// tamper node annotation
+	nodes, err := framework.Global.KubeClient.CoreV1().Nodes().List(context.TODO(),
+		metav1.ListOptions{LabelSelector: nc.WindowsOSLabel})
+	require.NoError(t, err)
+
+	for _, node := range nodes.Items {
+		patchData := fmt.Sprintf(`{"metadata":{"annotations":{"%s":"%s"}}}`, nc.VersionAnnotation, "badVersion")
+		_, err := framework.Global.KubeClient.CoreV1().Nodes().Patch(context.TODO(), node.Name, types.MergePatchType, []byte(patchData), metav1.PatchOptions{})
+		require.NoError(t, err)
+		if err == nil {
+			break
+		}
+	}
+
+	err = testCtx.waitForWindowsNodes(1, true, false, true)
+	if err != nil {
+		t.Fatal(t, "windows node upgrade failed  with %v", err)
+	}
+	// Get all the Windows nodes
+	nodes, err = framework.Global.KubeClient.CoreV1().Nodes().List(context.TODO(),
+		metav1.ListOptions{LabelSelector: nc.WindowsOSLabel})
+	require.NoError(t, err)
+
+	for _, node := range nodes.Items {
+		t.Run(node.GetName(), func(t *testing.T) {
+			require.Containsf(t, node.Annotations, nc.VersionAnnotation, "node missing version annotation")
+			assert.Equal(t, operatorVersion, node.Annotations[nc.VersionAnnotation], "Node recreation failed. WMCO version annotation mismatch")
+		})
+	}
+}
+
+// configureUpgradeTest carries out steps required before running tests for upgrade scenario.
+// The steps include -
+// 1. Scale down the operator to 0.
+// 2. Change Windows node version annotation to an invalid value
+// 3. Scale up the operator to 1
+func configureUpgradeTest() error {
+	// Scale down the WMCO deployment to 0
+	if err := scaleDeployment(0); err != nil {
+		return err
+	}
+
```

Comment:
To really simulate the upgrade scenario, you should wait for the deployment to scale down to 0.

Reply:
I am waiting for the deployment to scale up/down [here](https://github.com/openshift/windows-machine-config-operator/pull/101/files#diff-225c6dc1116ed17311dbd31aef8e09e9R147)

Reply:
But you are only checking `Spec.Replicas` there. That is going to take effect as soon as you patch deployment. Or am I missing something?

Reply:
Right. Thanks for catching that.